pkg/tfbridge/detailed_diff.go

```diff
@@ -0,0 +1,484 @@
+package tfbridge
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/contract"
+	pulumirpc "github.com/pulumi/pulumi/sdk/v3/proto/go"
+
+	"github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfbridge/info"
+	shim "github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfshim"
+)
+
+func isFlattened(s shim.Schema, ps *SchemaInfo) bool {
```

Comment:
I'd really love it if we didn't double-implement correlating shim.Schema and SchemaInfo here and delete this function. It is so error prone with MaxItems=1 and so forth.

I've introduced some path-based facilities in this code base that we can perhaps consider?

- type SchemaPath = walk.SchemaPath
- PropertyPathToSchemaPath
- func LookupSchemas(schemaPath SchemaPath, schemaMap shim.SchemaMap, schemaInfos map[string]*SchemaInfo) (shim.Schema, *SchemaInfo, error)

These things are perhaps a little inefficient compared to direct recursion (this can be fixed if we need to) but they centralize getting MaxItems=1 right. Would that work here? To use you track which level you're in using PropertyPath and then use these helpers to fetch `(shim.Schema, *SchemaInfo)`.